parser/src/aspects/struct.rs

```diff
@@ -0,0 +1,262 @@
+use lexer::token::TokenType;
+use lexer::token::TokenType::CurlyLeftBracket;
+
+use crate::aspects::expr_list::ExpressionListAspect;
+use crate::aspects::function_declaration::FunctionDeclarationAspect;
+use crate::aspects::r#type::TypeAspect;
+use crate::moves::{blanks, of_type, repeat, MoveOperations};
+use crate::parser::{ParseResult, Parser};
+
+pub trait StructAspect<'a> {
+    fn parse_struct(&mut self) -> ParseResult<StructDeclaration<'a>>;
+
+    fn parse_impl(&mut self) -> ParseResult<StructImpl<'a>>;
+}
+
+impl<'a> StructAspect<'a> for Parser<'a> {
+    fn parse_struct(&mut self) -> ParseResult<StructDeclaration<'a>> {
+        let start = self
+            .cursor
+            .force(of_type(TokenType::Struct), "`struct` expected")?;
+
+        let name = self.cursor.force(
+            blanks().then(of_type(TokenType::Identifier)),
+            "identifier expected",
+        )?;
+
+        self.cursor.advance(blanks());
+
+        let (parameters, _) = self.parse_explicit_list(
+            TokenType::SquaredLeftBracket,
+            TokenType::SquaredRightBracket,
+            Parser::parse_type_parameter,
+        )?;
+
+        self.cursor.advance(blanks());
+
+        let (fields, body_segment) = self.parse_explicit_list(
+            TokenType::CurlyLeftBracket,
+            TokenType::CurlyRightBracket,
+            Parser::parse_field,
+        )?;
+
+        let segment_start = self.cursor.relative_pos(start).start;
+        let segment_end = body_segment.end;
+        let segment = segment_start..segment_end;
+
+        Ok(StructDeclaration {
+            name: name.value,
+            parameters,
+            fields,
+            segment,
+        })
+    }
+
+    fn parse_impl(&mut self) -> ParseResult<StructImpl<'a>> {
+        let start = self
+            .cursor
+            .force(of_type(TokenType::Impl), "expected `struct`")?;
+
+        self.cursor.advance(blanks());
+
+        let (type_parameters, _) = self.parse_optional_list(
+            TokenType::SquaredLeftBracket,
+            TokenType::SquaredRightBracket,
+            Parser::parse_type_parameter,
+        )?;
+
+        self.cursor.advance(blanks());
+
+        let struct_type = self.parse_type()?;
+
+        self.cursor
+            .force(blanks().then(of_type(CurlyLeftBracket)), "expected `{`")?;
+
+        self.cursor.advance(blanks());
+
+        let mut functions = vec![];
+        let mut end_token = None;
+
+        while end_token.is_none() {
+            self.cursor
+                .advance(repeat(blanks().then(of_type(TokenType::SemiColon))));
+            self.cursor.advance(blanks());
+            functions.push(self.parse_function_declaration()?);
+
+            self.cursor
+                .advance(repeat(blanks().then(of_type(TokenType::SemiColon))));
+
+            end_token = self
+                .cursor
+                .advance(blanks().then(of_type(TokenType::CurlyRightBracket)));
+        }
```

Comment:
A leading semicolon is not accepted, and the syntax may not allow multiple semicolons.